### PR TITLE
feat(ws1): Add PyTorch matmul reference operator

### PR DESCRIPTION
**Pull request by frank-2077**
## Summary

  Adds the PyTorch reference GEMM/Matmul operator for Issue #108.

  This implements `NativeMatmulOp` as the fp32 ground-truth baseline for dense projection
  matmuls. The operator follows the frozen #108 interface contract:
  - `forward_fp32(a, b)` casts inputs to fp32 and calls `torch.matmul` once
  - `op_class = "reduction"`
  - registry dispatch via `kernel_registry.get_op("matmul")`

  Also adds operator documentation under `docs/operators/matmul.md`.

  ## Implementation

  - Added `rl_engine/kernels/ops/pytorch/linear/matmul.py`
  - Added `rl_engine/kernels/ops/pytorch/linear/__init__.py`
  - Registered `PYTORCH_NATIVE_MATMUL` in `rl_engine/kernels/registry.py`
  - Added `tests/test_matmul.py`
  - Added Matmul operator docs and linked them from `docs/operators/README.md`
  
  
image

## Summary by CodeRabbit

* **New Features**
  * Added a native PyTorch reference `matmul` operator.
  * Enhanced operator dispatch so the `matmul` implementation is selected consistently across CPU, CUDA, and ROCm.

* **Documentation**
  * Added a dedicated documentation page for the `matmul` operator.
  * Updated the operators index to include `matmul`.

* **Tests**
  * Added comprehensive `matmul` test coverage for correctness, dtype/FP32 behavior, supported shape variants, batch invariance, backward/gradient checks, and registry integration.

### CODE REVIEW

File: tests/test_matmul.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    def test_forward_vs_fp32_within_tolerance(self, dtype, atol, rtol):
+        op = NativeMatmulOp()
+        a, b = _make_inputs(2, 16, 64, 32, dtype=dtype)
+        out_typed = op.forward(a, b).float()
+        out_fp32 = op.forward_fp32(a, b)
+        diff = (out_typed - out_fp32).abs().max().item()
+        assert torch.allclose(out_typed, out_fp32, atol=atol, rtol=rtol), (
+            f"dtype={dtype}, max_abs_error={diff:.3e} exceeds " f"atol={atol}, rtol={rtol}"
+        )
+
```
Comment (Flink-ddd) [inline tests/test_matmul.py:176]: There are no backward pass tests (test_gradient_flows) in this PR. Matrix multiplication is the primary engine of the neural network and the most common source of gradient reduction drift.

You must verify that gradients flow correctly to both a (activations) and b (weights), and that they strictly adhere to the Axis-A Batch-Invariance contract:

For the activation gradient (a.grad), a slice of the full-batch gradient must be bitwise identical to the gradient computed from a sliced single-batch forward/backward pass.

For the weight gradient (b.grad), because it accumulates across the batch dimension, you should ideally verify that the full-batch weight gradient equals the sum of the single-batch weight gradients.

File: rl_engine/kernels/ops/pytorch/linear/matmul.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    It intentionally uses one `torch.matmul` call in fp32 for
+    the gold path and does not implement split-K or manual blocked accumulation.
+    """
+
+    op_class = "reduction"
+
+    def __init__(self) -> None:
+        pass
+
+    def __call__(self, a: Tensor, b: Tensor) -> Tensor:
```
Comment (Flink-ddd) [inline rl_engine/kernels/ops/pytorch/linear/matmul.py]: As with the previous WS1 reference operators, NativeMatmulOp must inherit from torch.nn.Module to ensure upstream compatibility with Dynamo tracing and PyTorch hooks.

Please inherit from nn.Module, initialize super().__init__(), and remove the manually defined __call__ method.

**Review by Flink-ddd — CHANGES_REQUESTED** [review]
Here are some review comments, everything else is fine.

File: tests/test_matmul.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    def test_batch_invariance_with_padding(self):
+        op = NativeMatmulOp()
+        a_valid, b = _make_inputs(2, 16, 64, 32, seed=456)
+        gen = torch.Generator().manual_seed(789)
+        padding = torch.randn(3, 16, 64, generator=gen)
+        a_padded = torch.cat([a_valid, padding], dim=0)
+        out_valid = op.forward_fp32(a_valid, b)
+        out_padded = op.forward_fp32(a_padded, b)
+        assert torch.equal(out_valid[0], out_padded[0])
+        assert torch.equal(out_valid[1], out_padded[1])
```
Comment (KJLdefeated) [inline tests/test_matmul.py:125]: cublas is not guarantee batch invariant, with larger size of matrix (M, K, N - Bigger K), batch-invariance may fail. Can you test on larger matrix?

**Review by KJLdefeated — CHANGES_REQUESTED** [review]

File: tests/test_matmul.py (hunk 3 above)
Comment (KJLdefeated) [reply tests/test_matmul.py:125]: You should alse add this to make sure batch invariant in CPU.
```python
@contextlib.contextmanager
def _single_thread():
    """Pin CPU GEMM to one thread so the matmul reduction order is batch-independent."""
    prev = torch.get_num_threads()
    torch.set_num_threads(1)
    try:
        yield
    finally:
        torch.set_num_threads(prev)
```

**Comment by Flink-ddd** [comment]
The Same, Please resolve the code conflicts first. If the review information has already resolved the issue or only requires explanation, please mark it. If there are no other issues, we will merge this PR. Thank you.

File: rl_engine/kernels/ops/pytorch/linear/matmul.py (hunk 2 above)
Comment (frank-2077) [reply rl_engine/kernels/ops/pytorch/linear/matmul.py]: Updated NativeMatmulOp to inherit torch.nn.Module and removed the manual __call__ override.

File: tests/test_matmul.py (hunk 1 above)
Comment (frank-2077) [reply tests/test_matmul.py:176]: Added backward coverage for NativeMatmulOp. The test checks gradients for both a and b against torch.matmul. 
· 65e2c23 Add native matmul backward test
Added batch-invariance coverage for backward. full_a.grad is checked bitwise against per-batch gradients, while full_b.grad is compared with assert_close  because b is shared across the batch and its gradient is a floating-point reduction over all per-batch contributions.
· 2a44ed7 Add native matmul batch grad invariance test

File: tests/test_matmul.py (hunk 3 above)
Comment (frank-2077) [reply tests/test_matmul.py:125]: Good point. I added single_thread in _2a44ed7_ commit. I added a backward test with K=512, N=384.

**Review by Flink-ddd — APPROVED** [review]
LGTM now, Thank you for update. cc @KJLdefeated PTAL again.

**Review by KJLdefeated — APPROVED** [review]
LTGM, thx!
